### PR TITLE
modification of take function so that it stays in PyArrow runtime

### PR DESCRIPTION
**Pull request by marc9595**
Modification of take function so that it stays in PyArrow runtime. This is done using code from ArrayInterface._to_numpy(). Take can now be called with list, arrays or slice.

### CODE REVIEW

File: fletcher/base.py [hunk 1]
```diff
@@ -703,28 +709,53 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
+            indices = self._indices_to_numpy_array(indices)
+            lengths = np.fromiter(map(len, self.data.chunks), dtype=np.int64)
+            cum_lengths = lengths.cumsum()
+
+            bins = np.searchsorted(cum_lengths, indices, side='right')  #one might want to parallel this operation
+            cum_lengths -= lengths
+            limits_idx = np.concatenate([[0], np.bincount(bins, minlength=self.data.num_chunks).cumsum()])
+
+            if is_increasing(bins):
+                sort_idx = None
```
Comment (artemru) [inline fletcher/base.py:737]: we don't need to sort anything if there's only one chunk, we can go directly for `pa.array.take`.

File: fletcher/base.py [hunk 2]
```diff
@@ -703,28 +709,53 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
         numpy.take
         pandas.api.extensions.take
         """
-        from pandas.core.algorithms import take
+        if isinstance(indices, list):
+            indices = np.asarray(indices)
+            if not np.any(indices < 0):
+                allow_fill = False
+        if isinstance(indices,slice) and allow_fill:
+            Raise Exception('allow_fill cannot be used with slice')
```
Comment (artemru) [inline fletcher/base.py]: python !!!
```suggestion
            raise ValueError('allow_fill cannot be used with slice')
```

**Comment by artemru** [comment]
in a case where we're going to take almost all indices from every chunk, it's going to be faster to make concat first and then take without sort/inverse_sort...

File: fletcher/base.py (hunk 1 above)
Comment (marc9595) [reply fletcher/base.py:737]: I have put a special case when there is only one chunk where `pa.array.take` is used.

**Comment by artemru** [comment]
@marc9595  could run tests on your local machine ?

File: fletcher/base.py [hunk 3]
```diff
@@ -664,14 +667,24 @@ def fillna(self, value=None, method=None, limit=None):
             new_values = self.copy()
         return new_values
 
-    def take(self, indices, allow_fill=False, fill_value=None):
+    def _indices_to_numpy_array(self, indices):
+        length = len(self)
+        if (isinstance(indices, slice) or np.any(indices < 0) or
+                (isinstance(indices, np.ndarray) and indices.dtype.kind == 'b')):
+            # pa.Array.take supports only positive indices
+            indices = np.arange(length)[indices]
```
Comment (artemru) [inline fletcher/base.py:675]: `dtype=np.int`. Btw, we could use `np.int32` if len(self) does not exceed int32

File: fletcher/base.py [hunk 4]
```diff
@@ -703,28 +709,56 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
-        it's called by :meth:`Series.reindex`, or any other method
-        that causes realignemnt, with a `fill_value`.
-
         See Also
         --------
         numpy.take
         pandas.api.extensions.take
         """
-        from pandas.core.algorithms import take
+        if isinstance(indices, list):
```
Comment (artemru) [inline fletcher/base.py:717]: why don't handle it in `_indices_to_numpy_array` ?

File: fletcher/base.py [hunk 5]
```diff
@@ -703,28 +709,56 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
+            indices = np.asarray(indices)
+            if not np.any(indices < 0):
+                allow_fill = False
+        if isinstance(indices,slice) and allow_fill:
```
Comment (artemru) [inline fletcher/base.py:721]: pep8

Comment (artemru) [reply fletcher/base.py:721]: (please activate a linter!)

File: fletcher/base.py [hunk 6]
```diff
@@ -703,28 +709,56 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
+            cum_lengths -= lengths
+            limits_idx = np.concatenate([[0], np.bincount(bins, minlength=self.data.num_chunks).cumsum()])
+
+            if is_increasing(bins):
+                sort_idx = None
+            else:
+                sort_idx = get_group_index_sorter(bins, self.data.num_chunks)
+                del bins
+                indices = indices[sort_idx]
+                sort_idx = np.argsort(sort_idx, kind="merge")  # inverse sort indices
```
Comment (artemru) [inline fletcher/base.py:742]: @krivonogov, I wonder if we could open-source our argsort from karma-perf ...

File: fletcher/base.py [hunk 7]
```diff
@@ -703,28 +709,56 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
-        # fill value should always be translated from the scalar
-        # type for the array, to the physical storage type for
-        # the data, before passing to take.
-        result = take(data, indices, fill_value=fill_value, allow_fill=allow_fill)
-        return self._from_sequence(result, dtype=self.data.type)
+            def take_in_one_chunk(i_chunk):
+                array_idx = indices[limits_idx[i_chunk]:limits_idx[i_chunk + 1]] - cum_lengths[i_chunk]
+                return self.data.chunk(i_chunk).take(pa.array(array_idx))#this is a pyarrow.Array
+
+            result = [take_in_one_chunk(i) for i in range(self.data.num_chunks)] #we know that self.data.num_chunks >1
```
Comment (artemru) [inline fletcher/base.py:748]: `result = list(map(...))`

File: fletcher/base.py [hunk 8]
```diff
@@ -703,28 +709,56 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
+
+        if not allow_fill:
+            indices = self._indices_to_numpy_array(indices)
+            if self.data.num_chunks==1:
+                return(FletcherArray(self.data.chunk(0).take(pa.array(indices))))
+
+            lengths = np.fromiter(map(len, self.data.chunks), dtype=np.int64)
+            cum_lengths = lengths.cumsum()
+
+            bins = np.searchsorted(cum_lengths, indices, side='right')  #one might want to parallel this operation
```
Comment (artemru) [inline fletcher/base.py:732]: given `bins` and `lengths` we could better chose an optimal path between `self.combine_array().take(...)` and the scenario below !

**Pull request closed without merging**